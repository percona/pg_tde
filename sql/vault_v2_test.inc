CREATE EXTENSION pg_tde;

\getenv root_token ROOT_TOKEN

SELECT pg_tde_add_key_provider_vault_v2('vault-incorrect',:'root_token','http://127.0.0.1:8200','DUMMY-TOKEN',NULL);
-- FAILS
SELECT pg_tde_set_principal_key('vault-v2-principal-key','vault-incorrect');

CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING :tde_am;

SELECT pg_tde_add_key_provider_vault_v2('vault-v2',:'root_token','http://127.0.0.1:8200','secret',NULL);
SELECT pg_tde_set_principal_key('vault-v2-principal-key','vault-v2');

CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING :tde_am;

INSERT INTO test_enc (k) VALUES (1);
INSERT INTO test_enc (k) VALUES (2);
INSERT INTO test_enc (k) VALUES (3);

SELECT * from test_enc;

DROP TABLE test_enc;

DROP EXTENSION pg_tde;
