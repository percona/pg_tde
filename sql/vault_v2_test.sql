CREATE EXTENSION pg_tde;

\getenv root_token_file VAULT_ROOT_TOKEN_FILE
\getenv cacert_file VAULT_CACERT_FILE

-- FAILS as mount path does not exist
SELECT pg_tde_add_database_key_provider_vault_v2('vault-incorrect', 'https://127.0.0.1:8200', 'DUMMY-MOUNT-PATH', :'root_token_file', :'cacert_file');

-- FAILS as it's not supported engine type
SELECT pg_tde_add_database_key_provider_vault_v2('vault-incorrect', 'https://127.0.0.1:8200', 'cubbyhole', :'root_token_file', :'cacert_file');

-- FAILS as it's not supported engine version
SELECT pg_tde_add_database_key_provider_vault_v2('vault-incorrect', 'https://127.0.0.1:8200', 'kv-v1', :'root_token_file', :'cacert_file');

SELECT pg_tde_add_database_key_provider_vault_v2('vault-v2', 'https://127.0.0.1:8200', 'secret', :'root_token_file', :'cacert_file');
SELECT pg_tde_create_key_using_database_key_provider('vault-v2-key', 'vault-v2');
SELECT pg_tde_set_key_using_database_key_provider('vault-v2-key', 'vault-v2');

CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING tde_heap;

INSERT INTO test_enc (k) VALUES (1);
INSERT INTO test_enc (k) VALUES (2);
INSERT INTO test_enc (k) VALUES (3);

SELECT * from test_enc;

SELECT pg_tde_verify_key();

DROP TABLE test_enc;

-- Creating provider fails if we can't connect to vault
SELECT pg_tde_add_database_key_provider_vault_v2('will-not-work', 'https://127.0.0.1:61', 'secret', :'root_token_file', :'cacert_file');

-- Changing provider fails if we can't connect to vault
SELECT pg_tde_change_database_key_provider_vault_v2('vault-v2', 'https://127.0.0.1:61', 'secret', :'root_token_file', :'cacert_file');

-- HTTPS without cert fails
SELECT pg_tde_change_database_key_provider_vault_v2('vault-v2', 'https://127.0.0.1:8200', 'secret', :'root_token_file', NULL);

-- HTTP against HTTPS server fails
SELECT pg_tde_change_database_key_provider_vault_v2('vault-v2', 'http://127.0.0.1:8200', 'secret', :'root_token_file', NULL);


-- Namespaces are supported, try using keys in a namespace. This test expect that the 'pgns' namespace exists, and it has a kv2 engine named 'secret'
SELECT pg_tde_add_database_key_provider_vault_v2('vault-v2ns', 'https://127.0.0.1:8200', 'secret', :'root_token_file', :'cacert_file', 'pgns');
SELECT pg_tde_create_key_using_database_key_provider('vault-v2-key-in-ns', 'vault-v2ns');
SELECT pg_tde_set_key_using_database_key_provider('vault-v2-key-in-ns', 'vault-v2ns');

CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING tde_heap;

INSERT INTO test_enc (k) VALUES (1);
INSERT INTO test_enc (k) VALUES (2);
INSERT INTO test_enc (k) VALUES (3);

SELECT * from test_enc;

SELECT pg_tde_verify_key();

DROP TABLE test_enc;

DROP EXTENSION pg_tde;
