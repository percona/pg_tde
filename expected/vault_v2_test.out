\set tde_am tde_heap
\i sql/vault_v2_test.inc
CREATE EXTENSION pg_tde;
\getenv root_token ROOT_TOKEN
SELECT pg_tde_add_key_provider_vault_v2('vault-incorrect',:'root_token','http://127.0.0.1:8200','DUMMY-TOKEN',NULL);
 pg_tde_add_key_provider_vault_v2 
----------------------------------
                                1
(1 row)

-- FAILS
SELECT pg_tde_set_principal_key('vault-v2-principal-key','vault-incorrect');
psql:sql/vault_v2_test.inc:7: ERROR:  Failed to store key on keyring. Please check the keyring configuration.
CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING :tde_am;
psql:sql/vault_v2_test.inc:13: ERROR:  failed to retrieve principal key. Create one using pg_tde_set_principal_key before using encrypted tables.
SELECT pg_tde_add_key_provider_vault_v2('vault-v2',:'root_token','http://127.0.0.1:8200','secret',NULL);
 pg_tde_add_key_provider_vault_v2 
----------------------------------
                                2
(1 row)

SELECT pg_tde_set_principal_key('vault-v2-principal-key','vault-v2');
 pg_tde_set_principal_key 
--------------------------
 t
(1 row)

CREATE TABLE test_enc(
	  id SERIAL,
	  k INTEGER DEFAULT '0' NOT NULL,
	  PRIMARY KEY (id)
	) USING :tde_am;
INSERT INTO test_enc (k) VALUES (1);
INSERT INTO test_enc (k) VALUES (2);
INSERT INTO test_enc (k) VALUES (3);
SELECT * from test_enc;
 id | k 
----+---
  1 | 1
  2 | 2
  3 | 3
(3 rows)

DROP TABLE test_enc;
DROP EXTENSION pg_tde;
